src/Compilers/CSharp/Portable/Lowering/ExtensionMethodReferenceRewriter.cs

```diff
@@ -0,0 +1,149 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Immutable;
+using System.Diagnostics;
+using Microsoft.CodeAnalysis.CSharp.Symbols;
+using Microsoft.CodeAnalysis.PooledObjects;
+
+namespace Microsoft.CodeAnalysis.CSharp
+{
+    /// <summary>
+    /// Replaces references to extension methods with references to their implementation methods
+    /// </summary>
+    internal class ExtensionMethodReferenceRewriter : BoundTreeRewriterWithStackGuardWithoutRecursionOnTheLeftOfBinaryOperator
+    {
+        private ExtensionMethodReferenceRewriter()
+        {
+        }
+
+        public static BoundStatement Rewrite(BoundStatement statement)
+        {
+            var rewriter = new ExtensionMethodReferenceRewriter();
+            return (BoundStatement)rewriter.Visit(statement);
+        }
+
+        public override BoundNode VisitCall(BoundCall node)
+        {
+            return VisitCall(this, node);
+        }
+
+        public static BoundNode VisitCall(BoundTreeRewriter rewriter, BoundCall node)
+        {
+            Debug.Assert(node != null);
+
+            BoundExpression rewrittenCall;
+
+            if (LocalRewriter.TryGetReceiver(node, out BoundCall? receiver1))
+            {
+                // Handle long call chain of both instance and extension method invocations.
+                var calls = ArrayBuilder<BoundCall>.GetInstance();
+
+                calls.Push(node);
+                node = receiver1;
+
+                while (LocalRewriter.TryGetReceiver(node, out BoundCall? receiver2))
+                {
+                    calls.Push(node);
+                    node = receiver2;
+                }
+
+                // Rewrite the receiver
+                BoundExpression? rewrittenReceiver = (BoundExpression?)rewriter.Visit(node.ReceiverOpt);
+
+                do
+                {
+                    rewrittenCall = visitArgumentsAndFinishRewrite(node, rewrittenReceiver);
+                    rewrittenReceiver = rewrittenCall;
+                }
+                while (calls.TryPop(out node!));
+
+                calls.Free();
+            }
+            else
+            {
+                // Rewrite the receiver
+                BoundExpression? rewrittenReceiver = (BoundExpression?)rewriter.Visit(node.ReceiverOpt);
+                rewrittenCall = visitArgumentsAndFinishRewrite(node, rewrittenReceiver);
+            }
+
+            return rewrittenCall;
+
+            BoundExpression visitArgumentsAndFinishRewrite(BoundCall node, BoundExpression? rewrittenReceiver)
```

Comment:
nit: consider passing `rewriter` explicitly and marking local function as `static` to avoid capture